server/oauth2.go

```diff
@@ -413,6 +413,18 @@ func (s *Server) parseAuthorizationRequest(r *http.Request) (*storage.AuthReques
 	scopes := strings.Fields(q.Get("scope"))
 	responseTypes := strings.Fields(q.Get("response_type"))
 
+	codeChallenge := q.Get("code_challenge")
+	codeChallengeMethod := q.Get("code_challenge_method")
+
+	if codeChallengeMethod == "" {
+		codeChallengeMethod = CodeChallengeMethodPlain
+	}
+
+	if codeChallengeMethod != CodeChallengeMethodS256 && codeChallengeMethod != CodeChallengeMethodPlain {
+		description := fmt.Sprintf("Unsupported PKCE challenge method (%q).", codeChallengeMethod)
+		return nil, &authErr{"", "", errInvalidRequest, description}
```

Comment:
I believe the error of unsupported PKCE challenge method should be returned as a query param in the redirect to a client callback URI.
So this check should be performed after redirectURI gets validated below and the newErr func defined there should be used.

Reply:
This is the passage from the RFC. I think that it is implemented according to it. What do you think?
```
RFC 7636                       OAUTH PKCE                 September 2015
4.4.1.  Error Response
   [...]
   If the server supporting PKCE does not support the requested
   transformation, the authorization endpoint MUST return the
   authorization error response with "error" value set to
   "invalid_request".  The "error_description" or the response of
   "error_uri" SHOULD explain the nature of error, e.g., transform
   algorithm not supported.
```

Reply:
The thing is, oauth2 rfc https://tools.ietf.org/html/rfc6749#section-4.2.2.1 says
```
If the request fails due to a missing, invalid, or mismatching
   redirection URI, or if the client identifier is missing or invalid,
   the authorization server SHOULD inform the resource owner of the
   error and MUST NOT automatically redirect the user-agent to the
   invalid redirection URI.

   **If the resource owner denies the access request or if the request
   fails for reasons other than a missing or invalid redirection URI,
   the authorization server informs the client by adding the following
   parameters to the fragment component of the redirection URI** using the
   "application/x-www-form-urlencoded" format, per Appendix B:
```

Using authErr{"", "", errInvalidRequest, description}  (with first two params empty) would result in dex just displaying page with error instead of sending HTTP redirect back to client

Reply:
Ok, now I understand. I will take a look at it.